### PR TITLE
Am/interceptors

### PR DESCRIPTION
**Pull request by admwx7**

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -3,12 +3,31 @@
...
   },
   "devDependencies": {
-    "iron-component-page": "PolymerElements/iron-component-page#^1.0.0",
-    "iron-demo-helpers": "PolymerElements/iron-demo-helpers#^1.0.0",
+    "iron-component-page": "PolymerElements/iron-component-page#^1.0 || ^2.0",
+    "iron-demo-helpers": "PolymerElements/iron-demo-helpers#^1.0 || ^2.0",
     "web-component-tester": "Polymer/web-component-tester#^6.0.0",
-    "webcomponentsjs": "webcomponents/webcomponentsjs#^0.7.0"
+    "webcomponentsjs": "webcomponents/webcomponentsjs#^1.0.0",
+    "iron-ajax": "^2.0.2"
```
Comment (stramel) [inline bower.json]: `PolymerElements/iron-ajax#^2.0.2`

Also, this is missing from 1.x variant.

File: demo/index.html [hunk 2]
```diff
@@ -10,7 +10,10 @@
 
     <link rel="import" href="../../iron-demo-helpers/demo-pages-shared-styles.html">
     <link rel="import" href="../../iron-demo-helpers/demo-snippet.html">
-    <link rel="import" href="../mtz-ajax-interceptor.html">
+    <link rel="import" href="mtz-ajax-interceptor.html">
+    <link rel="import" href="../mtz-status-code-interceptor.html">
+    <link rel="import" href="../mtz-presend-header-interceptor.html">
+    <link rel="import" href="../../iron-ajax/iron-ajax.html">
```
Comment (stramel) [inline demo/index.html]: NIT:  group component references at top.

File: demo/index.html [hunk 3]
```diff
@@ -20,7 +23,62 @@
...
+              </mtz-ajax-interceptor>
+              <p>
+                [[statusCodeIntercepted]]
+              </p>
+              <label>
+                Inject additional interceptors
+                <input type="checkbox" checked="{{headerInterceptor::change}}">
+              </label>
+              <br>
+              <button on-tap="intercept">Trigger Interceptors</button>
```
Comment (stramel) [inline demo/index.html]: use `on-click` since `on-tap` requires an extra import in 2.0.

File: demo/index.html [hunk 4]
```diff
@@ -20,7 +23,62 @@
...
+                    type: Boolean,
+                    value: false
+                  },
+                  statusCodeIntercepted: {
+                    type: String,
+                    value: ''
+                  }
+                },
+                listeners: {
+                  intercept: '_log'
```
Comment (stramel) [inline demo/index.html]: This won't transfer well to 2.0. `_log` is no longer available. Also, `listeners` object is going away.

File: demo/index.html [hunk 5]
```diff
@@ -20,7 +23,62 @@
...
+                },
+                intercept() {
+                  this.$.ajax.generateRequest();
+                },
+                statusCodeInterceptor() {
+                  this.statusCodeIntercepted = this.statusCodeIntercepted.concat('intercepted! ');
```
Comment (stramel) [inline demo/index.html]: any reason for a trailing space on the word?

File: mtz-ajax-interceptor-behavior.html [hunk 6]
```diff
@@ -1,13 +1,136 @@
...
+     *  then __registerEventListener is called.
+     * @private
+     *
+     * @param {CustomEvent} event
+     */
+    __registerInterceptor(event) {
+      const events = event.detail.events;
+      if (!events || !events.length) return;
+
+      event.stopPropagation();
```
Comment (stramel) [inline mtz-ajax-interceptor-behavior.html]: Move the stopPropagation to the first line. We still want to stop it even if they incorrectly sent the detail

File: mtz-ajax-interceptor-behavior.html [hunk 7]
```diff
@@ -1,13 +1,136 @@
...
+
+      // Element registered, now add an unregister listener
+      event.detail.addEventListener('unregister-interceptor', this.__unregisterInterceptor);
+
+      events.forEach((e) => {
```
Comment (stramel) [inline mtz-ajax-interceptor-behavior.html]: destructure for name?

File: mtz-ajax-interceptor-behavior.html [hunk 8]
```diff
@@ -1,13 +1,136 @@
...
+
+      // If propagation was stopped, ensure that the original logic is also called.
+      if (stopPropagation)
+        event.stopImmediatePropagation();
+    },
+    /**
+     * Registers an event listener for the provided event name against the runInterceptors method.
+     * @private
+     *
+     * @param {String} eventName
```
Comment (stramel) [inline mtz-ajax-interceptor-behavior.html:74]: Missing param description

File: mtz-ajax-interceptor-behavior.html [hunk 9]
```diff
@@ -1,13 +1,136 @@
...
+     */
+    __registerEventListener(eventName) {
+      this.addEventListener(eventName, this.runInterceptors);
+    },
+    /**
+     * Adds the interceptor stored in event.detail to each event referenced in events, if a new event group is added
+     *  then __registerEventListener is called.
+     * @private
+     *
+     * @param {CustomEvent} event
```
Comment (stramel) [inline mtz-ajax-interceptor-behavior.html]: Missing param description

File: mtz-ajax-interceptor-behavior.html [hunk 10]
```diff
@@ -1,13 +1,136 @@
...
+        this.interceptors[e.name].add(event.detail);
+      });
+
+      this.notifyPath('interceptors', this.interceptors);
+    },
+    /**
+     * Removes the event listener for the provided event name against the runInterceptors method.
+     * @private
+     *
+     * @param {String} eventName
```
Comment (stramel) [inline mtz-ajax-interceptor-behavior.html:110]: Missing param description

File: mtz-ajax-interceptor-behavior.html [hunk 11]
```diff
@@ -1,13 +1,136 @@
...
+     */
+    __unregisterEventListener(eventName) {
+      this.removeEventListener(eventName, this.runInterceptors);
+    },
+    /**
+     * Removes the interceptor from each event Set it's part of, if the set is emptied out then
+     *  __unregisterEventListener is called.
+     * @private
+     *
+     * @param {CustomEvent} event
```
Comment (stramel) [inline mtz-ajax-interceptor-behavior.html]: Missing param description

File: mtz-ajax-interceptor-behavior.html [hunk 12]
```diff
@@ -1,13 +1,136 @@
...
+     */
+    __unregisterInterceptor(event) {
+      const events = event.detail.events;
+      if (!events || !events.length) return;
+
+      events.forEach((e) => {
```
Comment (stramel) [inline mtz-ajax-interceptor-behavior.html]: Destructure for name

File: mtz-ajax-interceptor-behavior.html [hunk 13]
```diff
@@ -1,13 +1,136 @@
...
+    },
+    /**
+     * Removes the interceptor from each event Set it's part of, if the set is emptied out then
+     *  __unregisterEventListener is called.
+     * @private
+     *
+     * @param {CustomEvent} event
+     */
+    __unregisterInterceptor(event) {
+      const events = event.detail.events;
```
Comment (stramel) [inline mtz-ajax-interceptor-behavior.html:124]: stopPropagation?

File: mtz-auth-interceptor.html [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+        value: String
+      },
+      /**
+       * Checks the event to see if event.target has with-auth set, if so then _applyHeader is called.
+       *
+       * @param {CustomEvent} event - the event that triggered the interceptor
+       * @param {String} eventName - the name of the event that triggered this intercept
+       */
+      intercept(event, eventName) {
+        console.log('test', event);
```
Comment (stramel) [inline mtz-auth-interceptor.html]: Remove console.log

File: mtz-auth-interceptor.html [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+            ];
+          }
+        },
+        /* The header name to attach to the ajax request */
+        key: {
+          type: String,
+          value: 'Authorization'
+        },
+        /* The value to associate with the provided header */
+        value: String
```
Comment (stramel) [inline mtz-auth-interceptor.html:35]: Hm, I wonder if `token` or `authToken` would be better for auth interceptor.

File: mtz-status-code-interceptor.html [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+          type: Array,
+          value() {
+            return [
+              {
+                name: 'response',
+                path: 'detail.status'
+              }, {
+                name: 'upload-response',
+                // TODO: is this right?
+                path: 'detail.status'
```
Comment (stramel) [inline mtz-status-code-interceptor.html]: `detail.xhr.status`

File: mtz-status-code-interceptor.html [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+       * @param {String} eventName - the event name that triggered the intercept
+       * @param {Number} [status = this.status] - the status code to check against
+       */
+      intercept(event, eventName, status = this.status) {
+        let statusCode;
+        const path = (this.events.find((item) => item.name === eventName) || {}).path;
+
+        if (path)
+          statusCode = this.get(path, event);
+        if (Number(status) === Number(statusCode))
```
Comment (stramel) [inline mtz-status-code-interceptor.html]: Should this be nested? `statusCode` won't get set if path isn't defined.

File: test/mtz-ajax-interceptor_test.html [hunk 18]
```diff
@@ -9,34 +9,265 @@
...
-    </dom-module>
-
     <test-fixture id="basic">
       <template>
         <mtz-ajax-interceptor></mtz-ajax-interceptor>
       </template>
     </test-fixture>
 
     <script>
+      // TODO: add tests
```
Comment (stramel) [inline test/mtz-ajax-interceptor_test.html]: Probably don't need this anymore.

File: test/mtz-ajax-interceptor_test.html [hunk 19]
```diff
@@ -9,34 +9,265 @@
...
+                }
+              ]),
+            };
+            element.runInterceptors('foo', event);
+            const interceptors = Array.from(element.interceptors.foo);
+            expect(interceptors[0].intercept).calledWithMatch(event, 'foo');
+            expect(interceptors[1].intercept).calledWithMatch(event, 'foo');
+            expect(interceptors[2].intercept).calledWithMatch(event, 'foo');
+            expect(interceptors[1].intercept).calledBefore(interceptors[2].intercept);
+            expect(interceptors[2].intercept).calledBefore(interceptors[0].intercept);
```
Comment (stramel) [inline test/mtz-ajax-interceptor_test.html]: Should we also check that 1 is called after 0?

File: all-imports.html [hunk 20]
```diff
@@ -1 +1 @@
-<link rel="import" href="mtz-ajax-interceptor-behavior.html">
+<link rel="import" href="mtz-status-code-interceptor.html">
```
Comment (stramel) [inline all-imports.html]: Don't see the point in your all-imports file if it doesn't have *all* imports....

**Review by stramel — COMMENTED** [review]
Looks pretty good overall

File: demo/index.html (hunk 3 above)
Comment (haydenmcp) [reply demo/index.html]: @stramel Doesn't on-tap also operate for mobile devices, though? It may be good for the demo to be accessible from mobile.

File: mtz-ajax-interceptor-behavior.html [hunk 21]
```diff
@@ -1,13 +1,136 @@
...
--->
+  /**
+   * `mtz.AjaxInterceptorBehavior` handles registering declarative interceptors then listening for all registered events
+   *  and calling the interceptors associated with those events.
+   *
+   * @polymerBehavior mtz.AjaxInterceptorBehavior
+   */
+  mtz.AjaxInterceptorBehavior = {
+    properties: {
+      /* The map of event to interceptor Set to listen against */
```
Comment (haydenmcp) [inline mtz-ajax-interceptor-behavior.html]: NIT: The wording of this comment is a bit confusing, especially taken with the word "Set" starting with a cap. Revise to be more clear. Maybe just "Event to interceptor set map"?

File: mtz-ajax-interceptor-behavior.html [hunk 22]
```diff
@@ -1,13 +1,136 @@
...
+    },
+    /**
+     * Adds the interceptor stored in event.detail to each event referenced in events, if a new event group is added
+     *  then __registerEventListener is called.
+     * @private
+     *
+     * @param {CustomEvent} event
+     */
+    __registerInterceptor(event) {
+      const events = event.detail.events;
```
Comment (haydenmcp) [inline mtz-ajax-interceptor-behavior.html]: Is there a need for a check on event.detail.events being defined? This may output an error message in the dev tools console otherwise.

File: mtz-ajax-interceptor-behavior.html [hunk 23]
```diff
@@ -1,13 +1,136 @@
...
+      /* The map of event to interceptor Set to listen against */
+      interceptors: {
+        type: Object,
+        value() {
+          return {};
+        }
+      }
+    },
+    attached() {
+      this.addEventListener('register-interceptor', this.__registerInterceptor);
```
Comment (haydenmcp) [inline mtz-ajax-interceptor-behavior.html:24]: For attached lifecycle callback the docs specify that hybrid elements use Polymer.RenderStatus.afterNextRender to attach listeners to hybrid elements. This is snippet from docs:

For work that can be deferred until after first paint (such as adding event listeners), you can use Polymer.RenderStatus.afterNextRender, which takes the same arguments as beforeNextRender.

You've probably already seen it, but if not check out: https://www.polymer-project.org/2.0/docs/upgrade

File: mtz-ajax-interceptor-behavior.html [hunk 24]
```diff
@@ -1,13 +1,136 @@
...
-<script>
-  window.mtz = mtz || {};
+      // Override stopImmediatePropagation so we can break out if needed.
+      const stopImmediatePropagationFnc = event.stopImmediatePropagation;
+      let stopPropagation = false;
+      event.stopImmediatePropagation = () => {
+        stopPropagation = true;
+      };
+
+      for (const interceptor of Array.from(interceptors).sort((a, b) => b.priority - a.priority)) {
```
Comment (haydenmcp) [inline mtz-ajax-interceptor-behavior.html]: This fcn in really well written. Very, very easy to read. + 1

File: demo/index.html (hunk 3 above)
Comment (stramel) [reply demo/index.html]: @haydenmcp it does also operate on mobile devices. I would say for now on hybrid we use on-click for ease. We can look back into it for 2.0.

Comment (stramel) [reply demo/index.html]: I'm not really wanting to have to create 1.x and 2.0 demos separately.

File: mtz-ajax-interceptor-behavior.html (hunk 21 above)
Comment (stramel) [reply mtz-ajax-interceptor-behavior.html]: @haydenmcp Set is the type. Maybe use of ticks around would make more sense?

File: mtz-ajax-interceptor-behavior.html (hunk 23 above)
Comment (stramel) [reply mtz-ajax-interceptor-behavior.html:24]: Generally, I think we will want to. I'm not sure on this we would want to because these are critical to making requests.

File: mtz-ajax-interceptor-behavior.html (hunk 21 above)
Comment (haydenmcp) [reply mtz-ajax-interceptor-behavior.html]: I got that after review, but still think it's too wordy. It's nitpicky. But I think just "event to interceptor set map" makes more immediate sense. lol

File: mtz-interceptor-behavior.html [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+    },
+    /**
+     * Used to fire the event that will trigger custom logic for this interceptor.
+     * @protected
+     * @fires mtz.InterceptorBehavior#intercept
+     *
+     * @param {CustomEvent} event - the captured event that triggered this interceptor.
+     */
+    _fireInterceptEvent(event) {
+      const e = new CustomEvent('intercept', {
```
Comment (stramel) [inline mtz-interceptor-behavior.html]: Should we just put this in `dispatchEvent` instead?

File: bower.json (hunk 1 above)
Comment (admwx7) [reply bower.json]: Both 1.x and mainline need ^2.0 since that's where the presend event was added.

Comment (stramel) [reply bower.json]: That's fine, I was just pointing out that it was missing from the 1.x and it wasn't using the github repository.

File: demo/index.html (hunk 5 above)
Comment (admwx7) [reply demo/index.html]: This is a visual representation going into the DOM via a binding, was easier to read if there was a space between each entry added since spamming it would end up with "intercepted! intercepted! ..."

File: mtz-ajax-interceptor-behavior.html (hunk 23 above)
Comment (admwx7) [reply mtz-ajax-interceptor-behavior.html:24]: I'm with Stramel on this one, there is no DOM to render from this element, all of it's functionality it hinged upon those event listeners, and when combined with something like `<iron-ajax auto>` and when the interceptors fire their registration event, deferring to after the next render may be too late.

File: mtz-ajax-interceptor-behavior.html (hunk 8 above)
Comment (admwx7) [reply mtz-ajax-interceptor-behavior.html:74]: Description is an optional portion of the `@param` annotation in JSDocs, and I always exclude them if the name of the param tells you everything you need to know about the param, in this case adding "the name of the event" adds nothing more here.  Same as below, adding "the event that triggered this event listener" doesn't provide any more details then having a function that listens for an event then naming your param "event"

Comment (stramel) [reply mtz-ajax-interceptor-behavior.html:74]: I'm pretty sure that eslint will yell if it is left out... may be wrong though.

File: mtz-auth-interceptor.html (hunk 15 above)
Comment (admwx7) [reply mtz-auth-interceptor.html:35]: I'm trying to match it to the presend-header as close as possible so it can easily be an extension in 2.0

File: test/mtz-ajax-interceptor_test.html (hunk 19 above)
Comment (admwx7) [reply test/mtz-ajax-interceptor_test.html]: Order: 1 => 2 => 0
Checks:
* 1 => 2
* 2 => 0

The only one not explicitly checked is 1 => 0, but that's implied since 2 is there.

File: bower.json (hunk 1 above)
Comment (admwx7) [reply bower.json]: Do I need to explicitly add it to both?  I thought that the 1.x inherited mainline if it didn't override?

File: mtz-ajax-interceptor-behavior.html (hunk 8 above)
Comment (admwx7) [reply mtz-ajax-interceptor-behavior.html:74]: I haven't had problems and I've been running it without up to this point.

File: mtz-ajax-interceptor-behavior.html (hunk 13 above)
Comment (admwx7) [reply mtz-ajax-interceptor-behavior.html:124]: Don't have to, this is that weird case we talked about where it's detached from the DOM already so you have to grab the element reference and explicitly attach it, which means it could never bubble even if it wanted to, however if multiple elements are explicitly doing that process I don't want to block any propagation.

File: test/mtz-ajax-interceptor_test.html (hunk 19 above)
Comment (stramel) [reply test/mtz-ajax-interceptor_test.html]: But you are checking after all have run, so 1 could still technically be called before 2  but after 0.

File: mtz-ajax-interceptor-behavior.html (hunk 8 above)
Comment (stramel) [reply mtz-ajax-interceptor-behavior.html:74]: Also, you gave a description for eventName above but not here. at least stay consistent.

File: mtz-ajax-interceptor-behavior.html (hunk 13 above)
Comment (stramel) [reply mtz-ajax-interceptor-behavior.html:124]: oh yeah. no problem.

File: mtz-auth-interceptor.html [hunk 26]
```diff
@@ -0,0 +1,59 @@
+<link rel="import" href="../polymer/polymer.html">
+<link rel="import" href="mtz-interceptor-behavior.html">
+
+<!--
+`mtz-auth-interceptor`
+An implementation of the mtz.InterceptorBehavior that injects auth headers based on the event.target containing the
+  [with-auth] attribute.
+-->
```
Comment (stramel) [inline mtz-auth-interceptor.html]: Add `@demo demo/index.html` annotation

File: mtz-presend-header-interceptor.html [hunk 27]
```diff
@@ -0,0 +1,54 @@
+<link rel="import" href="../polymer/polymer.html">
+<link rel="import" href="mtz-interceptor-behavior.html">
+
+<!--
+`mtz-presend-header-interceptor`
+An implementation of the mtz.InterceptorBehavior that injects a header:value onto a presend event for iron-ajax.
+-->
```
Comment (stramel) [inline mtz-presend-header-interceptor.html]: Add `@demo demo/index.html` annotation

File: mtz-status-code-interceptor.html [hunk 28]
```diff
@@ -0,0 +1,54 @@
+<link rel="import" href="../polymer/polymer.html">
+<link rel="import" href="mtz-interceptor-behavior.html">
+
+<!--
+`mtz-status-code-interceptor`
+An implementation of the mtz.InterceptorBehavior that compares a status property against the status code in the event
+to determine if the interceptor logic should run or not.
+-->
```
Comment (stramel) [inline mtz-status-code-interceptor.html]: Add `@demo demo/index.html` annotation

**Review by stramel — APPROVED** [review]

**Review by haydenmcp — APPROVED** [review]
This is really good code. Awesome job. Really easy to read, really easy to follow. We may want to use this as an example for onboarding.
